### PR TITLE
workspaces: react to Reconfigure

### LINKED ISSUE #764: React to changes in workspaces on Reconfigure
```
19:48:56 <Flrian> what exactly was >workspaces not reacting to Reconfigure
19:48:56 <Flrian> ?
19:49:17 <Consolatis> you redefine workspaces in your rc.xml and Reconfigure
19:49:23 <Consolatis> e.g. runtime changes of the workspaces
19:49:49 <Flrian> ahh
19:50:06 <Flrian> so ideally they should
19:50:08 <Consolatis> its not implemented and you have to restart labwc so the new workspace config is actually used
19:50:10 <Consolatis> yes
19:50:44 <Consolatis> its a bit tricky to move views from no-more existing workspaces somewhere and stuffs but its not that hard to implement
```

### PR DESCRIPTION
**Pull request by tokyo4j**
Addresses #764

### CODE REVIEW

**Comment by johanmalm** [comment]
Nice. Can look through tonight if nobody else gets there first.

File: src/workspaces.c [hunk 1]
```diff
@@ -388,15 +388,79 @@ workspaces_find(struct workspace *anchor, const char *name, bool wrap)
...
+workspaces_reconfigure(struct server *server)
+{
+	/*
+	 * Currently, we consider updates like below as "WS2 is renamed to WS3
+	 * and WS2 is added":
+	 * old workspaces = ["WS1", "WS2"]
+	 * new workspaces = ["WS1", "WS3", "WS2"]
+	 */
+	struct workspace *conf;
+	struct wl_list *ws_link = server->workspaces.next;
```
Comment (johanmalm) [inline src/workspaces.c:410]: Consider `s/ws_link/elm/`

The actual `wl_list` macros/functions use `elm` in places. We've also done the same in a few instances.
https://github.com/nobled/wayland/blob/f5ccd7a9e670e097ff9b5faae8a26b3af4bb7b60/src/wayland-util.h#L112

File: src/workspaces.c [hunk 2]
```diff
@@ -388,15 +388,79 @@ workspaces_find(struct workspace *anchor, const char *name, bool wrap)
...
+	wl_list_for_each(conf, &rc.workspace_config.workspaces, link) {
+		if (ws_link == &server->workspaces) {
+			/* # of workspaces increased */
```
Comment (johanmalm) [inline src/workspaces.c:413]: Suggest adding the work "configured" before workspaces here.

File: src/workspaces.c [hunk 3]
```diff
@@ -388,15 +388,79 @@ workspaces_find(struct workspace *anchor, const char *name, bool wrap)
...
+			wlr_log(WLR_DEBUG, "Adding workspace \"%s\"", conf->name);
+			add_workspace(server, conf->name);
+			continue;
+		}
+		struct workspace *ws = wl_container_of(ws_link, ws, link);
```
Comment (johanmalm) [inline src/workspaces.c:418]: Consider `s/ws/actual_workspace/`

File: src/workspaces.c [hunk 4]
```diff
@@ -388,15 +388,79 @@ workspaces_find(struct workspace *anchor, const char *name, bool wrap)
...
+			ws->name = xstrdup(conf->name);
+		}
+		ws_link = ws_link->next;
+	}
+
+	if (ws_link == &server->workspaces) {
+		return;
+	}
+
+	/* # of workspaces decreased */
```
Comment (johanmalm) [inline src/workspaces.c:433]: Consider adding the work `configured` here.

File: src/workspaces.c [hunk 5]
```diff
@@ -388,15 +388,79 @@ workspaces_find(struct workspace *anchor, const char *name, bool wrap)
...
+	overlay_hide(&server->seat);
+	struct workspace *first_ws =
+		wl_container_of(server->workspaces.next, first_ws, link);
+	while (ws_link != &server->workspaces) {
+		struct workspace *ws = wl_container_of(ws_link, ws, link);
```
Comment (johanmalm) [inline src/workspaces.c:438]: Again. Considering `s/ws/actual_workspace/`

**Review by johanmalm — COMMENTED** [review]
LGTM. Really nice patch. Have tested and it works very well.

I've made some comments about variable names. This is always tricky/subjective so feel free to merge without adopting changes.

Reading through it took me a while to navigate what was the `configured` workspace versus the `actual` (already existing) workspace.

File: src/workspaces.c (hunk 5 above)
Comment (tokyo4j) [reply src/workspaces.c]: Done.

File: src/workspaces.c (hunk 4 above)
Comment (tokyo4j) [reply src/workspaces.c]: Done.

File: src/workspaces.c (hunk 3 above)
Comment (tokyo4j) [reply src/workspaces.c]: Done.

File: src/workspaces.c (hunk 2 above)
Comment (tokyo4j) [reply src/workspaces.c]: Done.

File: src/workspaces.c (hunk 1 above)
Comment (tokyo4j) [reply src/workspaces.c]: I don't think `elm` is a good name here.
`elm` seems fine if it is used only in `for(;;)` or `wl_container_of(...)`, but it is used throughout this function here and `elm` lacks its context.

I replaced it with `actual_workspace_link` instead.

File: src/workspaces.c [hunk 6]
```diff
@@ -388,15 +388,79 @@ workspaces_find(struct workspace *anchor, const char *name, bool wrap)
...
+void
+workspaces_reconfigure(struct server *server)
+{
+	/*
+	 * Currently, we consider updates like below as "WS2 is renamed to WS3
+	 * and WS2 is added":
+	 * old workspaces = ["WS1", "WS2"]
+	 * new workspaces = ["WS1", "WS3", "WS2"]
+	 */
+	struct workspace *conf;
```
[the remark this answers is not in the document]
Comment (tokyo4j) [reply src/workspaces.c]: Done.

File: src/workspaces.c [hunk 7]
```diff
@@ -388,15 +388,79 @@ workspaces_find(struct workspace *anchor, const char *name, bool wrap)
...
+	zfree(workspace->name);
+	wl_list_remove(&workspace->link);
+	free(workspace);
+}
+
+void
+workspaces_reconfigure(struct server *server)
+{
+	/*
+	 * Currently, we consider updates like below as "WS2 is renamed to WS3
```
[the remark this answers is not in the document]
Comment (tokyo4j) [reply src/workspaces.c]: It now looks like:
``` 
- Update names
- Add workspaces if more workspaces are desired 
- Destroy workspaces if fewer workspace are desired
```

**Comment by tokyo4j** [comment]
@johanmalm
Merge this PR if you're okay with it.
Any more fix suggestions are fine and I can fix and merge it myself then.

**Comment by johanmalm** [comment]
LGTM.

Happy with all changes.

Sorry to add one more... I committed a change this afternoon which lets us get rid of the `g_` in `workspaces_reconfigure(g_server);` Suggest a rebase against `master` and removal of that `g_`.

**Comment by tokyo4j** [comment]
> Sorry to add one more... I committed a change this afternoon which lets us get rid of the g_ in workspaces_reconfigure(g_server); Suggest a rebase against master and removal of that g_ before merge.

I'm just wondering: If I clicked "Merge pull request" button without rebase, is a commit that cannot be built created in the master? Or another CI runs?

**Comment by johanmalm** [comment]
Good stuff.

> I'm just wondering: If I clicked "Merge pull request" button without rebase, is a commit that cannot be built created in the master? Or another CI runs?

Looks like you've locally rebased and then force-pushed (which is what I also would have done). Perfect.

Out of the three GitHub "merge options"

1. `Create a merge commit`
2. `Squash and merge`
3. `Rebase and merge`

...we normally go for `Rebase and merge` (which it looks like you've done).

If there is a long PR thread with useful information and a commit history with more than one commit that we want to preserve, then we sometimes do `Create a merge commit` so that we have a record of the PR number - but it's quite unusual.

**Comment by Consolatis** [comment]
> I'm just wondering: If I clicked "Merge pull request" button without rebase, is a commit that cannot be built created in the master? Or another CI runs?

If you would have `g_server` in this PR and then merge it then the master branch would indeed not build.

The CI (or more specifically the [build.yml](https://github.com/labwc/labwc/blob/master/.github/workflows/build.yml) part of it) in this repo is configured to run in PRs only because the master branch is protected and can not be pushed to directly so there is not really a point in running it on pushes. We do run the update for https://labwc.github.io on pushes to master though if some `docs/*.scd` file was changed due to a merged PR.

**Comment by tokyo4j** [comment]
Thank you all! It was really valuable experience!
> The CI (or more specifically the [build.yml](https://github.com/labwc/labwc/blob/master/.github/workflows/build.yml) part of it) in this repo is configured to run in PRs only because the master branch is protected and can not be pushed to directly so there is not really a point in running it on pushes.

**Comment by tokyo4j** [comment]
Thank you all!
So it seems that I need to careful to merge a PR if other PRs are merged after the latest push on the PR.

> Squash and merge can also be useful
> - to quickly change a commit subject or body without having to force-push to the PR branch
> - or if dealing with somebody who is not that experienced with git so there are multiple fixup commits around and squashing them from the maintainer side is faster and less error prone than asking the person from the PR to do it

Good to know. I wondered how you sometimes changes the commit subject.

> Edit 2:
> If you add Fixes: #764 to your commit message, GH will automatically close the issue when merging the PR.

I knew it works with PR description, but didn't know it works also for commit message. Thanks.

**Comment by Consolatis** [comment]
> So it seems that I need to careful to merge a PR if other PRs are merged after the latest push on the PR.

Hm.. I guess we could make the build jobs rebase on master before compiling. Or have a another Arch build named `Rebase` that does it so its easier to see if only that job fails.

**Pull request merged**